Add scroll factor input command.

sxmichaels opened the pull request:
This commit adds a `scroll_factor` input command whose value is used as a multiplier for axis event deltas received from a device. Put simply, this means that scrolling speed can now be adjusted, a configuration feature that libinput itself lacks. At least on my laptop's touchpad, such a change is sorely needed.

For example, adding `scroll_factor 0.5` under a device's `input` section will make scrolling half as fast.

Fixes #3004.

RyanDwyer commented on sway/commands/input/scroll_factor.c:
```diff
@@ -0,0 +1,26 @@
+	float scroll_factor = atof(argv[0]);
+	if (scroll_factor < 0) {
+		return cmd_results_new(CMD_INVALID, "scroll_factor",
+			"Input out of range [0, inf)");
```
The brackets aren't matching. Although I'd prefer wording such as "Scroll factor cannot be negative".

emersion replied:
I think it's deliberate, it's a math notation. But yeah natural language seems more appropriate here.

RyanDwyer commented:
Can you also document it in the `sway-input` man page?

sxmichaels commented:
OK, implemented all of @emersion 's fixes and documented the new option in the sway-input man page. One question: is checking `errno != 0` sufficient, or should I provide separate checks and messages for each possible type of error `strtof` can encounter, e.g. out-of-range & underflow?

emersion commented on sway/input/cursor.c:
```diff
@@ -1025,7 +1029,8 @@ static void dispatch_cursor_axis(struct sway_cursor *cursor,
 	wlr_seat_pointer_notify_axis(cursor->seat->wlr_seat, event->time_msec,
-		event->orientation, event->delta, event->delta_discrete, event->source);
+		event->orientation, scroll_factor * event->delta,
+		scroll_factor * event->delta_discrete, event->source);
```
`event->delta_discrete` is an integer, so maybe it should be rounded instead of being silently floored.

emersion commented on sway/sway-input.5.scd:
```diff
@@ -105,14 +105,18 @@ The following commands may only be used in the configuration file.
 	can be obtained from `libinput debug-events`.
 	If set to 0, it disables the scroll\_button on\_button\_down.
 
+*input* <identifier> scroll\_factor <floating point value >= 0>
```
That `>= 0` is strange here, Can you instead add a sentence in the description?

emersion commented on sway/commands/input/scroll_factor.c:
```diff
@@ -0,0 +1,30 @@
+			"scroll_factor", "No input device defined.");
+	}
+
+	char *end;
```
Can you set `errno = 0` right before calling `strtof`? Otherwise it might fail because of a previous `errno` value.

emersion commented:
>is checking errno != 0 sufficient, or should I provide separate checks and messages for each possible type of error strtof can encounter, e.g. out-of-range & underflow?

I think just checking `errno` is enough.

sxmichaels commented:
Great! Thanks for the help.

sxmichaels commented:
Er, hold on — it looks like something is wrong with the parsing of `scroll_factor` now. To anyone who might merge this: don't yet.

ascent12 commented on common/util.c:
```diff
@@ -139,6 +140,17 @@ bool parse_boolean(const char *boolean, bool current) {
+	float flt = strtof(value, &end);
+	if (*end || errno) {
+		wlr_log(WLR_DEBUG, "Invalid float value %s, defaulting to FLT_MIN", value);
+		return FLT_MIN;
```
NAN would be a better sentinel value than this.
Or change the signature to `bool parse_float(const char *value, float *ret)`.

sxmichaels commented:
Okay, not sure what was going on there; I may not have compiled the latest version when I last tested. Everything works now. I also realized that `pointer_accel.c` was also using the `atof` method that I had used in the first commit of this PR, so I moved the float parsing code into a `parse_float` function in `common/util.c` and changed both `scroll_factor.c` and `pointer_accel.c` to use it. It takes a `const char *` and returns `FLT_MIN` on error.

emersion commented:
Ping

sxmichaels commented:
Oh dear, very sorry about the delay — somehow I didn't see your last message. I just incorporated your changes and rebased. This was a particularly ugly rebase due to a bunch of merge conflicts in `util.h` and `util.c` … looks like the final result is right, but I seem to have created a bunch of garbage intermediate commits in the process. I wasn't able to figure out a way around this.

RedSoxFan commented on sway/input/cursor.c:
```diff
@@ -992,6 +995,7 @@ static void dispatch_cursor_axis(struct sway_cursor *cursor,
 	enum wlr_edges edge = cont ? find_edge(cont, cursor) : WLR_EDGE_NONE;
 	bool on_border = edge != WLR_EDGE_NONE;
 	bool on_titlebar = cont && !on_border && !surface;
+	float scroll_factor = (ic == NULL || ic->scroll_factor == FLT_MIN) ? 1.0f : ic->scroll_factor;
```
Style: Lines should not exceed 80 columns

ddevault commented on ISSUE_TEMPLATE.md:
```diff
@@ -1,4 +1,11 @@
+
+If `lsmod | grep nvidia | wc -l` shows anything other than zero, your bug report is not welcome here.
+
+Otherwise, please include the following four components in your bug report: sway version, debug log, configuration (if applicable), and an explanation of steps taken to reproduce the issue.
```
I think you botched a merge/rebase/something

sxmichaels commented:
OK, I figured I might have to do that. Still not sure what the heck happened during rebasing, so I just completely redid the changes as a single comment as @RedSoxFan suggested.

RedSoxFan commented on sway/commands/input/scroll_factor.c:
```diff
@@ -0,0 +1,32 @@
+	}
+
+	float scroll_factor = parse_float(argv[0]);
+	if (scroll_factor == NAN) {
```
Sorry. I missed this before. You cannot compare a floating point value with `NAN` to determine whether the value is `NAN` since `NAN != NAN`. You can either use `isnan(scroll_factor)` or `scroll_factor != scroll_factor`.

sxmichaels commented:
Changed the two instances of `x == NAN` to `isnan(x)` (there was also one in `pointer_accel.c`) and rebased. This rebase went OK.

sxmichaels commented:
Oh dear. Having done that, all the commits to master since I created this PR become a part of the PR itself, which is the same problem I ran into when I first tried to rebase. I'm stuck at this point; never seen this happen before. Is there some git setting that is making my rebases behave differently than would normally be expected?

RedSoxFan commented:
Can you please post the output of `git remote -v`? I'll give you instructions after that. Hopefully, it will help you figure out what you are doing differently.

sxmichaels commented:
Sure.

```
origin    git@github.com:sxmichaels/sway.git (fetch)
origin    git@github.com:sxmichaels/sway.git (push)
upstream    https://github.com/swaywm/sway (fetch)
upstream    https://github.com/swaywm/sway (push)
```

RedSoxFan commented:
Try the following:
```
git fetch upstream
git checkout add-scroll-factor
git rebase upstream/master
git push -f origin add-scroll-factor
```

emersion commented:
It seems it's down to one commit again somehow.

RedSoxFan commented:
Looks like that worked. It's rebased now.

sxmichaels commented:
Ah! _That_ did it this time, I think, although strangely I could have sworn I'd tried that before. Out of curiosity, do you know what I was doing wrong when rebasing in the first place? I'm still not exactly sure.

Thanks very much for the help — and sorry for all this trouble for such a simple PR!

RedSoxFan commented:
> Out of curiosity, do you know what I was doing wrong when rebasing in the first place? 

I think either your forks `master` or what you had fetched for `upstream/master` were outdated and on a commit that was reverted. I'm not sure which since it looks like your fork is up-to-date now. Just make sure to do `git fetch upstream` before `git rebase upstream/master` (if you just do `git rebase master`, it will rebase against `origin/master` -- which is your fork).

RedSoxFan commented:
Thanks!

mvdan commented:
Thanks so much for this change! Two years later, but I just found out about it and it really made an external mouse much easier to use :)
